File: bluetooth/src/main/java/com/ankerboxmanager/bluetooth/impl/AnkerBoxBluetoothLe.java
//package com.ankerboxmanager.bluetooth.impl;
//
//import static com.ankerboxmanager.bluetooth.impl.AnkerBoxBluetoothHelper.setNotify;
//import static com.ankerboxmanager.bluetooth.impl.AnkerBoxBluetoothHelper.uuidFromString;
//
//import java.util.ArrayList;
//import java.util.List;
//import java.util.concurrent.ExecutorService;
//import java.util.concurrent.Executors;
//
//import com.ankerboxmanager.bluetooth.IBluetoothLe;
//import com.ankerboxmanager.bluetooth.bean.AnkerBoxBluetoothGattCharacteristic;
//import com.ankerboxmanager.bluetooth.bean.AnkerBoxBluetoothLeService;
//import com.ankerboxmanager.bluetooth.scaner.BluetoothScanner;
//import com.ankerboxmanager.bluetooth.scaner.LegacyLeBluetoothScanner;
//import com.ankerboxmanager.bluetooth.scaner.LollipopLeBluetoothScanner;
//
//import android.annotation.TargetApi;
//import android.bluetooth.BluetoothDevice;
//import android.bluetooth.BluetoothGatt;
//import android.bluetooth.BluetoothGattCharacteristic;
//import android.bluetooth.BluetoothGattService;
//import android.content.Context;
//import android.content.pm.PackageManager;
//import android.os.Build;
//import android.util.Base64;
//import android.util.Log;
//
///**
// * Created on 21/03/2018.
// *
// * @author lx
// */
//
//@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
//public class AnkerBoxBluetoothLe extends AnkerBoxBluetooth implements IBluetoothLe {
//
//    private boolean isSupportLe;
//
//    private BluetoothGatt mBluetoothGatt;
//
//    private ExecutorService executorService = Executors.newFixedThreadPool(5);
//
//    private GattCallbackImpl gattCallback;
//
//    @Override
//    public void init(Context context) {
//        super.init(context);
//        Log.d(TAG, "le init");
//        gattCallback = new GattCallbackImpl();
//        isSupportLe = isSupportBluetooth() && context.getPackageManager().hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE);
//    }
//
//    @Override
//    protected BluetoothScanner createBluetoothScanner(Context context) {
//        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
//            return new LollipopLeBluetoothScanner(context);
//        } else {
//            return new LegacyLeBluetoothScanner(context);
//        }
//    }
//
//    @Override
//    public void destroy(Context context) {
//        super.destroy(context);
//        if (gattCallback != null) {
//            gattCallback.destroy();
//            gattCallback = null;
//        }
//        if (mBluetoothGatt != null) {
//            mBluetoothGatt.disconnect();
//            mBluetoothGatt.close();
//            mBluetoothGatt = null;
//        }
//        if (executorService != null) {
//            executorService.shutdown();
//            executorService = null;
//        }
//        Log.d(TAG, "le destroy");
//    }
//
//    @Override
//    public boolean isSupportLe() {
//        return isSupportLe;
//    }
//
//    @Override
//    public void startDevicesDiscovery(String[] serviceUUIDs, AsyncReturnCallback returnCallback) {
//        if (isSupportLe()) {
//            super.startDevicesDiscovery(serviceUUIDs, returnCallback);
//        } else {
//            if (returnCallback != null) {
//                returnCallback.onReturn(false);
//            }
//        }
//    }
//
//    @Override
//    public void stopDevicesDiscovery(AsyncReturnCallback returnCallback) {
//        if (isSupportLe()) {
//            super.stopDevicesDiscovery(returnCallback);
//        } else {
//            if (returnCallback != null) {
//                returnCallback.onReturn(false);
//            }
//        }
//    }
//
//    @Override
//    public void connectGatt(final Context context, final String deviceId, final AsyncReturnCallback returnCallback) {
//        Log.d(TAG, "connectGatt, deviceId=" + deviceId);
//        if (isSupportLe()) {
//            executorService.submit(new Runnable() {
//                @Override
//                public void run() {
//                    BluetoothDevice bondedDevice = mBluetoothAdapter.getRemoteDevice(deviceId);
//                    mBluetoothGatt = bondedDevice.connectGatt(context, true, gattCallback);
//                    gattCallback.setGattConnect(mBluetoothGatt, returnCallback);
//                }
//            });
//        }
//    }
//
//    @Override
//    public boolean disconnectGatt(Context context, String deviceId) {
//        Log.d(TAG, "disconnectGatt, deviceId=" + deviceId);
//        if (mBluetoothGatt != null) {
//            mBluetoothGatt.disconnect();
//            mBluetoothGatt.close();
//            mBluetoothGatt = null;
//            return true;
//        }
//        return false;
//    }
//
//    @Override
//    public boolean addLeStateChangedListener(StateChangedListener listener) {
//        Log.d(TAG, "addLeStateChangedListener, listener=" + listener);
//        return gattCallback.addLeStateChangedListener(listener);
//    }
//
//    @Override
//    public boolean removeLeStateChangedListener(StateChangedListener listener) {
//        Log.d(TAG, "removeLeStateChangedListener, listener=" + listener);
//        return gattCallback.removeLeStateChangedListener(listener);
//    }
//
//    @Override
//    public List<AnkerBoxBluetoothLeService> getServices(Context context, String deviceId) {
//        Log.d(TAG, "getServices, deviceId = " + deviceId);
//        if (mBluetoothGatt != null) {
//            List<AnkerBoxBluetoothLeService> services = new ArrayList<>();
//            List<BluetoothGattService> gattServices = mBluetoothGatt.getServices();
//            Log.d(TAG, "gattServices.size = " + gattServices.size());
//            for (BluetoothGattService gattService : gattServices) {
//                AnkerBoxBluetoothLeService service = new AnkerBoxBluetoothLeService();
//                service.uuid = gattService.getUuid();
//                service.isPrimary = gattService.getType() == BluetoothGattService.SERVICE_TYPE_PRIMARY;
//                Log.d(TAG, "service.uuid = " + service.uuid + "service.isPrimary = " + service.isPrimary);
//                services.add(service);
//            }
//            return services;
//        }
//        return null;
//    }
//
//    @Override
//    public List<AnkerBoxBluetoothGattCharacteristic> getDeviceCharacteristics(Context context, String deviceId,
//                                                                              String serviceId) {
//        Log.d(TAG, "getDeviceCharacteristics deviceId= " + deviceId + ", serviceId = " + serviceId);
//        if (mBluetoothGatt != null) {
//            List<AnkerBoxBluetoothGattCharacteristic> characteristicList = new ArrayList<>();
//            BluetoothGattService gattService = mBluetoothGatt.getService(uuidFromString(serviceId));
//            List<BluetoothGattCharacteristic> gattCharacteristics = gattService.getCharacteristics();
//            Log.d(TAG, "gattCharacteristics.size = " + gattCharacteristics.size());
//            for (BluetoothGattCharacteristic gattCharacteristic : gattCharacteristics) {
//                AnkerBoxBluetoothGattCharacteristic characteristic = new AnkerBoxBluetoothGattCharacteristic();
//                characteristic.uuid = gattCharacteristic.getUuid();
//                int properties = gattCharacteristic.getProperties();
//                characteristic.characteristic = new AnkerBoxBluetoothGattCharacteristic.Characteristic();
//                characteristic.characteristic.read = properties & BluetoothGattCharacteristic.PROPERTY_READ;
//                characteristic.characteristic.write = (properties & BluetoothGattCharacteristic.PROPERTY_WRITE)
//                        | (properties & BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE) ;
//                characteristic.characteristic.notify = properties & BluetoothGattCharacteristic.PROPERTY_NOTIFY;
//                characteristic.characteristic.indicate = properties & BluetoothGattCharacteristic.PROPERTY_INDICATE;
//                characteristicList.add(characteristic);
//                Log.d(TAG, characteristic.toString());
//            }
//            return characteristicList;
//        }
//        return null;
//    }
//
//    @Override
//    public boolean readCharacteristicValue(Context context, String deviceId, String serviceId, String characteristicId) {
//        Log.d(TAG, "readCharacteristicValue deviceId= " + deviceId + ", serviceId = " + serviceId + ", characteristicId = " + characteristicId);
//        if (mBluetoothGatt != null) {
//            BluetoothGattService gattService = mBluetoothGatt.getService(uuidFromString(serviceId));
//            BluetoothGattCharacteristic characteristic = gattService.getCharacteristic(uuidFromString(characteristicId));
//            return mBluetoothGatt.readCharacteristic(characteristic);
//        }
//        return false;
//    }
//
//    @Override
//    public boolean writeCharacteristicValue(Context context, String deviceId, String serviceId, String characteristicId,
//                                            String value) {
//        Log.d(TAG, "writeCharacteristicValue deviceId= " + deviceId + ", serviceId = " + serviceId
//                + ", characteristicId = " + characteristicId + ", value = " + value);
//        if (mBluetoothGatt != null) {
//            BluetoothGattService gattService = mBluetoothGatt.getService(uuidFromString(serviceId));
//            BluetoothGattCharacteristic characteristic = gattService.getCharacteristic(uuidFromString(characteristicId));
//            characteristic.setValue(Base64.decode(value, Base64.DEFAULT));
//            characteristic.setWriteType(BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT);
//            return mBluetoothGatt.writeCharacteristic(characteristic);
//        }
//        return false;
//    }
//
//    @Override
//    public boolean notifyCharacteristicValueChange(Context context, String deviceId, String serviceId,
//                                                   String characteristicId, int state) {
//        Log.d(TAG, "notifyCharacteristicValueChange deviceId= " + deviceId + ", serviceId = " + serviceId
//                + ", characteristicId = " + characteristicId + ", state = " + state);
//        if (mBluetoothGatt != null) {
//           return setNotify(mBluetoothGatt, uuidFromString(serviceId), uuidFromString(characteristicId), state == 1);
//        }
//        return false;
//    }
//
//    @Override
//    public boolean addCharacteristicValueChangedListener(CharacteristicValueChangedListener listener) {
//        Log.d(TAG, "addCharacteristicValueChangedListener, listener=" + listener);
//        return gattCallback.addCharacteristicValueChangedListener(listener);
//    }
//
//    @Override
//    public boolean removeCharacteristicValueChangedListener(CharacteristicValueChangedListener listener) {
//        Log.d(TAG, "removeCharacteristicValueChangedListener, listener=" + listener);
//        return gattCallback.removeCharacteristicValueChangedListener(listener);
//    }
//
//
//
//}

File: bluetooth/src/main/kotlin/com/ankerboxmanger/bluetooth/impl/AnkerBoxBluetoothLe.kt
package com.ankerboxmanger.bluetooth.impl

import android.annotation.TargetApi
import com.ankerboxmanger.bluetooth.impl.AnkerBoxBluetoothHelper.setNotify
import com.ankerboxmanger.bluetooth.impl.AnkerBoxBluetoothHelper.uuidFromString

import java.util.ArrayList
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattService
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.util.Base64
import android.util.Log
import com.ankerboxmanger.bluetooth.ext.notNull
import com.ankerboxmanger.bluetooth.ext.safeLet
import com.ankerboxmanger.bluetooth.IBluetoothLe
import com.ankerboxmanger.bluetooth.bean.AnkerBoxBluetoothGattCharacteristic
import com.ankerboxmanger.bluetooth.bean.AnkerBoxBluetoothLeService
import com.ankerboxmanger.bluetooth.scaner.BluetoothScanner
import com.ankerboxmanger.bluetooth.scaner.LegacyLeBluetoothScanner
import com.ankerboxmanger.bluetooth.scaner.LollipopLeBluetoothScanner
import com.ankerboxmanger.bluetooth.IBluetooth
import com.ankerboxmanger.bluetooth.IBluetooth.Companion.TAG

/**
 * Created on 21/03/2018.
 *
 * @author lx
 */

@TargetApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
class AnkerBoxBluetoothLe : AnkerBoxBluetooth(), IBluetoothLe {

    override var isSupportLe: Boolean = false
        private set

    private var mBluetoothGatt: BluetoothGatt? = null

    private var executorService: ExecutorService? = Executors.newFixedThreadPool(5)

    private var gattCallback: GattCallbackImpl? = null

    override fun init(context: Context) {
        super.init(context)
        Log.d(TAG, "le init")
        gattCallback = GattCallbackImpl()
        isSupportLe = isSupportBluetooth && context.packageManager.hasSystemFeature(PackageManager.FEATURE_BLUETOOTH_LE)
    }

    override fun createBluetoothScanner(context: Context): BluetoothScanner {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            LollipopLeBluetoothScanner(context)
        } else {
            LegacyLeBluetoothScanner(context)
        }
    }

    override fun destroy(context: Context) {
        super.destroy(context)
        gattCallback?.destroy()
        gattCallback = null
        mBluetoothGatt?.disconnect()
        mBluetoothGatt?.close()
        mBluetoothGatt = null
        executorService?.shutdown()
        executorService = null
        Log.d(TAG, "le destroy")
    }

    override fun startDevicesDiscovery(serviceUUIDs: Array<String>, returnCallback: IBluetooth.AsyncReturnCallback?) {
        if (isSupportLe) {
            super.startDevicesDiscovery(serviceUUIDs, returnCallback)
        } else {
            returnCallback?.onReturn(false)
        }
    }

    override fun stopDevicesDiscovery(returnCallback: IBluetooth.AsyncReturnCallback?) {
        if (isSupportLe) {
            super.stopDevicesDiscovery(returnCallback)
        } else {
            returnCallback?.onReturn(false)
        }
    }

    override fun connectGatt(context: Context, deviceId: String, returnCallback: IBluetooth.AsyncReturnCallback) {
        Log.d(TAG, "connectGatt, deviceId=$deviceId")
        if (isSupportLe) {
            safeLet(executorService, mBluetoothAdapter) { service, adapter ->
                service.execute {
                    val bondedDevice = adapter.getRemoteDevice(deviceId)
                    val gatt = bondedDevice.connectGatt(context, true, gattCallback)
                    gattCallback?.setGattConnect(gatt, returnCallback)
                    mBluetoothGatt = gatt
                }
            }
        }
    }

    override fun disconnectGatt(context: Context, deviceId: String): Boolean {
        Log.d(TAG, "disconnectGatt, deviceId=$deviceId")
        if (mBluetoothGatt != null) {
            mBluetoothGatt!!.disconnect()
            mBluetoothGatt!!.close()
            mBluetoothGatt = null
            return true
        }
        return false
    }

    override fun addLeStateChangedListener(listener: IBluetoothLe.StateChangedListener): Boolean {
        Log.d(TAG, "addLeStateChangedListener, listener=$listener")
        return gattCallback.let { it != null && addLeStateChangedListener(listener) }
    }

    override fun removeLeStateChangedListener(listener: IBluetoothLe.StateChangedListener): Boolean {
        Log.d(TAG, "removeLeStateChangedListener, listener=$listener")
        return gattCallback.let { it != null && removeLeStateChangedListener(listener) }
    }

    override fun getServices(context: Context, deviceId: String): List<AnkerBoxBluetoothLeService>? {
        Log.d(TAG, "getServices, deviceId = $deviceId")
        mBluetoothGatt.notNull {
            val services = ArrayList<AnkerBoxBluetoothLeService>()
            Log.d(TAG, "gattServices.size = " + it.services.size)
            for (gattService in it.services) {
                val service = AnkerBoxBluetoothLeService(gattService.uuid, gattService.type == BluetoothGattService.SERVICE_TYPE_PRIMARY)
                Log.d(TAG, "service.uuid = ${service.uuid}, service.isPrimary = ${service.isPrimary}")
                services.add(service)
            }
            return services
        }
        return null
    }

    override fun getDeviceCharacteristics(context: Context, deviceId: String,
                                          serviceId: String): List<AnkerBoxBluetoothGattCharacteristic>? {
        Log.d(TAG, "getDeviceCharacteristics deviceId= $deviceId, serviceId = $serviceId")
        mBluetoothGatt.notNull {
            val characteristicList = ArrayList<AnkerBoxBluetoothGattCharacteristic>()
            val gattService = it.getService(uuidFromString(serviceId))
            val gattCharacteristics = gattService.characteristics
            Log.d(TAG, "gattCharacteristics.size = ${gattCharacteristics.size}")
            for (gattCharacteristic in gattCharacteristics) {
                val properties = gattCharacteristic.properties
                val read = properties and BluetoothGattCharacteristic.PROPERTY_READ
                val write = properties and BluetoothGattCharacteristic.PROPERTY_WRITE or
                        (properties and BluetoothGattCharacteristic.PROPERTY_WRITE_NO_RESPONSE)
                val notify = properties and BluetoothGattCharacteristic.PROPERTY_NOTIFY
                val indicate = properties and BluetoothGattCharacteristic.PROPERTY_INDICATE
                val characteristic = AnkerBoxBluetoothGattCharacteristic.Characteristic(read, write, notify, indicate)
                characteristicList.add(AnkerBoxBluetoothGattCharacteristic(gattCharacteristic.uuid, characteristic))
                Log.d(TAG, characteristic.toString())
            }
            return characteristicList
        }
        return null
    }

    override fun readCharacteristicValue(context: Context, deviceId: String, serviceId: String, characteristicId: String): Boolean {
        Log.d(TAG, "readCharacteristicValue deviceId= $deviceId, serviceId = $serviceId, characteristicId = $characteristicId")
        mBluetoothGatt.notNull {
            val gattService = it.getService(uuidFromString(serviceId))
            val characteristic = gattService.getCharacteristic(uuidFromString(characteristicId))
            return it.readCharacteristic(characteristic)
        }
        return false
    }

    override fun writeCharacteristicValue(context: Context, deviceId: String, serviceId: String, characteristicId: String,
                                          value: String): Boolean {
        Log.d(TAG, "writeCharacteristicValue deviceId= " + deviceId + ", serviceId = " + serviceId
                + ", characteristicId = " + characteristicId + ", value = " + value)
        mBluetoothGatt.notNull {
            val gattService = it.getService(uuidFromString(serviceId))
            val characteristic = gattService.getCharacteristic(uuidFromString(characteristicId))
            characteristic.value = Base64.decode(value, Base64.DEFAULT)
            characteristic.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
            return it.writeCharacteristic(characteristic)
        }
        return false
    }

    override fun notifyCharacteristicValueChange(context: Context, deviceId: String, serviceId: String,
                                                 characteristicId: String, state: Int): Boolean {
        Log.d(TAG, "notifyCharacteristicValueChange deviceId= " + deviceId + ", serviceId = " + serviceId
                + ", characteristicId = " + characteristicId + ", state = " + state)
        return mBluetoothGatt.let {
            it != null && setNotify(it, uuidFromString(serviceId), uuidFromString(characteristicId), state == 1)
        }
    }

    override fun addCharacteristicValueChangedListener(listener: IBluetoothLe.CharacteristicValueChangedListener): Boolean {
        Log.d(TAG, "addCharacteristicValueChangedListener, listener=$listener")
        return gattCallback.let { it != null && addCharacteristicValueChangedListener(listener) }
    }

    override fun removeCharacteristicValueChangedListener(listener: IBluetoothLe.CharacteristicValueChangedListener): Boolean {
        Log.d(TAG, "removeCharacteristicValueChangedListener, listener=$listener")
        return gattCallback.let { it != null && removeCharacteristicValueChangedListener(listener) }
    }

}
